### PR TITLE
exception when not anonymous and no user/pass

### PR DESCRIPTION
**Pull request by jmbeach**
When the OPC UA server does not have anonymous authentication enabled and no username and password are provided, you get the following error:

> System.ArgumentNullException: 'String reference not set to an instance of a String.
> Parameter name: s'

at line 592 of UaTcpSessionChannel.cs

```
var userNameIdentity = (UserNameIdentity)this.UserIdentity;
byte[] passwordBytes = System.Text.Encoding.UTF8.GetBytes(userNameIdentity.Password);
```

because the password is null.

This change throws an exception that is more clear. Let me know if I should change the Exception message or Exception class.

Thanks

### CODE REVIEW

File: UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs [hunk 1]
```diff
@@ -589,6 +589,11 @@ protected override async Task OnOpenAsync(CancellationToken token = default(Canc
                 }
 
                 var userNameIdentity = (UserNameIdentity)this.UserIdentity;
+                if (userNameIdentity.UserName == null && userNameIdentity.Password == null)
```
Comment (quinmars) [inline UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs]: Shouldn't that be ORed, i.e., `userNameIdentity.UserName == null || userNameIdentity.Password == null`?

File: UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs [hunk 2]
```diff
@@ -589,6 +589,11 @@ protected override async Task OnOpenAsync(CancellationToken token = default(Canc
...
+                {
+                    throw new Exception("Authentication is not anonymous, but no username and password were provided.");
```
Comment (quinmars) [inline UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs]: A plain `Exception` is very general. I think an `InvalidOperationException` could work here. According to its [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.invalidoperationexception?view=netframework-4.8):

> The exception that is thrown when a method call is invalid for the object's current state.

File: UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs (hunk 1 above)
Comment (jmbeach) [reply UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs]: Wasn't sure. Is it possible to have authentication with a username and no password?

Comment (jmbeach) [reply UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs]: Well I guess you would always want to check the password, or else you'll get the exception so yeah, I'll `or` it

File: UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs (hunk 2 above)
Comment (jmbeach) [reply UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs]: Changed to `InvalidOperationException`

File: UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs (hunk 1 above)
Comment (quinmars) [reply UaClient/ServiceModel/Ua/Channels/UaTcpSessionChannel.cs]: That was my thought, too.

**Comment by awcullen** [comment]
I took a look at couple of other clients and conclude it's best to try to send the username token and get back a service exception from the server if the token is not valid. So the clients do something like:
```
byte[] passwordBytes = userNameIdentity.Password != null ? System.Text.Encoding.UTF8.GetBytes(userNameIdentity.Password): new byte[0];
```

**Comment by jmbeach** [comment]
what do you think of my latest commit as a compromise. I tried removing the exception first and just using the empty byte array if no username and password are provided, but the exception is very vague:

> The user identity token is not valid

and it's one of those things you might spend a decent amount of time tracking down.

I prefer user identity token not valid to the exception

> System.ArgumentNullException: 'String reference not set to an instance of a String.
Parameter name: s'

**Comment by awcullen** [comment]
I agree with you (and appreciate your help).  Its just that I recently participated in OPC Foundation Compliance Testing for another product.  When OPCF tests clients and servers it uses automated tools and runs many hundreds of tests on every conceivable request and response. For instance it tests every variation of username and password, null and not null,  and validates the server response looking for the exact status code it expects.  
One day I'd like this package to pass Compliance Testing,  so I think we better let the generic server response stay.  Would you have a place to put this check for nulls in your application code, where you ask the user for name and password?

**Comment by quinmars** [comment]
I think the point of @jmbeach is that the `null` password will never reach the server because `System.Text.Encoding.UTF8.GetBytes(null)` will throw an exception before.

**Comment by awcullen** [comment]
Oh, I do agree that null password should be handled as so:
```
byte[] passwordBytes = userNameIdentity.Password != null ? System.Text.Encoding.UTF8.GetBytes(userNameIdentity.Password): new byte[0];
```
Just that a null username should be passed to the server like normal.  The Compliance Test will expect the server to respond 'The user identity token is not valid'

**Comment by jmbeach** [comment]
I think it's a great idea to meet the compliance testing. My latest commit only has the null check in it.

Let me try to describe the issue I was trying to address a little bit better: I normally use Kepware as my OPC server for new projects and they semi-recently made a change that all new projects default to having anonymous authentication disabled. I don't get a specific error, but nothing is working.

Obviously now that I've spent this much time thinking about the issue, I probably won't run into it again, but others might.

The null check will now at least give me an error to point me in the direction of authentication. Do you think it would make sense to put something in the Readme that says
> if you get the error "The user identity token is not valid", check your OPC UA server's authentication configuration

?

Either way, thanks for your time and reviewing this change.

**Comment by awcullen** [comment]
Thanks for fixing this bug!

**Comment by awcullen** [comment]
@jmbeach You should take a look at the sample 'ConsoleApp' https://github.com/convertersystems/opc-ua-samples/blob/2d050938d0c8684ce4be164a0a1f2d9c8704cd18/ConsoleApp/Program.cs#L174
or the wpf sample 'StatusHmi'.

If the anonymous Identity is disabled then the client knows to ask for a username and password.
Give it a try, and let me know.

**Pull request merged**